cpp/src/arrow/ipc/reader.h

```diff
@@ -190,6 +190,15 @@ class ARROW_EXPORT RecordBatchFileReader
   /// \return the read batch
   virtual Result<std::shared_ptr<RecordBatch>> ReadRecordBatch(int i) = 0;
 
+  /// \brief Read a particular record batch along with its custom metadatda from the file.
+  /// Does not copy memory if the input source supports zero-copy.
+  ///
+  /// \param[in] i the index of the record batch to return
+  /// \return a pair of the read batch and its custom metadata
+  virtual Result<
+      std::pair<std::shared_ptr<RecordBatch>, std::shared_ptr<KeyValueMetadata>>>
+  ReadRecordBatchWithCustomMetadata(int i) = 0;
```

Comment:
This is a new API added for the `RecordBatchFileReader` class so that accessing a record batch with its metadata  by the index is possible.

Reply:
I think user code would be more readable with:
```c++
struct RecordBatchWithMetadata {
  std::shared_ptr<RecordBatch> batch;
  std::shared_ptr<KeyValueMetadata> custom_metadata;
};
virtual Result<RecordBatchWithMetadata> ReadRecordBatchWithMetadata(int i) = 0;
```

Reply:
Sure. I change the code to return the struct now.